Server connection doesn't support TLS 1.3

<!--- Provide a general summary of the issue in the Title above -->

## Description

Here is the exception from the logcat:

```
12-19 11:15:35.792 17092 17092 W System.err: javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb400006fd4119cd8: Failure in SSL library, usually a protocol error
12-19 11:15:35.792 17092 17092 W System.err: error:1000042e:SSL routines:OPENSSL_internal:TLSV1_ALERT_PROTOCOL_VERSION (external/boringssl/src/ssl/tls_record.cc:486 0xb4000071541c1f60:0x00000003)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.SSLUtils.toSSLHandshakeException(SSLUtils.java:358)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.convertException(ConscryptEngine.java:1131)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:916)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:744)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:709)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:907)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.-$$Nest$mprocessDataFromSocket(Unknown Source:0)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:242)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:224)
12-19 11:15:35.793 17092 17092 W System.err: 	at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:102)
12-19 11:15:35.793 17092 17092 W System.err: 	at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:220)
12-19 11:15:35.793 17092 17092 W System.err: 	at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
12-19 11:15:35.793 17092 17092 W System.err: 	at io.grpc.okhttp.OkHttpClientTransport$3.run(OkHttpClientTransport.java:739)
12-19 11:15:35.793 17092 17092 W System.err: 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
12-19 11:15:35.793 17092 17092 W System.err: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1154)
12-19 11:15:35.793 17092 17092 W System.err: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:652)
12-19 11:15:35.793 17092 17092 W System.err: 	at java.lang.Thread.run(Thread.java:1563)
12-19 11:15:35.793 17092 17092 W System.err: Caused by: javax.net.ssl.SSLProtocolException: Read error: ssl=0xb400006fd4119cd8: Failure in SSL library, usually a protocol error
12-19 11:15:35.793 17092 17092 W System.err: error:1000042e:SSL routines:OPENSSL_internal:TLSV1_ALERT_PROTOCOL_VERSION (external/boringssl/src/ssl/tls_record.cc:486 0xb4000071541c1f60:0x00000003)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.NativeSsl.readDirectByteBuffer(NativeSsl.java:571)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.readPlaintextDataDirect(ConscryptEngine.java:1092)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1076)
12-19 11:15:35.793 17092 17092 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:873)
12-19 11:15:35.793 17092 17092 W System.err: 	... 14 more
```

## Expected Behavior

Connect with TLS 1.3 if available.

## Actual Behavior

Connects with TLS 1.2, or fails if server forces TLS 1.3.

## Steps to Reproduce

Try to connect to a server that forces TLS 1.3.

## Your Environment

- BitBanana version: 1.0.0
- Android version: 16
- Device: Pixel 7a
- LND Version: 0.20.0-beta (behind Traefik 3.6.5 which terminates TLS)

Tested with 1.0.1 and it can't connect to TLS 1.3. Same exception I guess:

```
04-14 11:35:18.340  1280  1280 W System.err: javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073ebe80ad8: Failure in SSL library, usually a protocol error
04-14 11:35:18.341  1280  1280 W System.err: error:1000042e:SSL routines:OPENSSL_internal:TLSV1_ALERT_PROTOCOL_VERSION (external/boringssl/src/ssl/tls_record.cc:486 0xb40000739be9b9b0:0x00000003)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.SSLUtils.toSSLHandshakeException(SSLUtils.java:358)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.convertException(ConscryptEngine.java:1131)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:916)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:744)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:709)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:912)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.-$$Nest$mprocessDataFromSocket(Unknown Source:0)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:242)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:224)
04-14 11:35:18.341  1280  1280 W System.err: 	at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:102)
04-14 11:35:18.341  1280  1280 W System.err: 	at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:220)
04-14 11:35:18.341  1280  1280 W System.err: 	at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
04-14 11:35:18.341  1280  1280 W System.err: 	at io.grpc.okhttp.OkHttpClientTransport$3.run(OkHttpClientTransport.java:739)
04-14 11:35:18.341  1280  1280 W System.err: 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
04-14 11:35:18.341  1280  1280 W System.err: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1154)
04-14 11:35:18.341  1280  1280 W System.err: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:652)
04-14 11:35:18.341  1280  1280 W System.err: 	at java.lang.Thread.run(Thread.java:1564)
04-14 11:35:18.341  1280  1280 W System.err: Caused by: javax.net.ssl.SSLProtocolException: Read error: ssl=0xb4000073ebe80ad8: Failure in SSL library, usually a protocol error
04-14 11:35:18.341  1280  1280 W System.err: error:1000042e:SSL routines:OPENSSL_internal:TLSV1_ALERT_PROTOCOL_VERSION (external/boringssl/src/ssl/tls_record.cc:486 0xb40000739be9b9b0:0x00000003)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.NativeSsl.readDirectByteBuffer(NativeSsl.java:571)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.readPlaintextDataDirect(ConscryptEngine.java:1092)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1076)
04-14 11:35:18.341  1280  1280 W System.err: 	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:873)
04-14 11:35:18.341  1280  1280 W System.err: 	... 14 more
```

--- a/app/src/main/java/app/michaelwuensch/bitbanana/backends/coreLightning/connection/CoreLightningConnection.java
+++ b/app/src/main/java/app/michaelwuensch/bitbanana/backends/coreLightning/connection/CoreLightningConnection.java
@@ -88,6 +88,7 @@ public class CoreLightningConnection {
                         .forAddress(host, port)
                         .hostnameVerifier(hostnameVerifier) // null = default hostnameVerifier
                         .sslSocketFactory(inspectingFactory) // null = default SSLSocketFactory
+                        .tlsConnectionSpec(RefConstants.TLS_VERSIONS, RefConstants.CIPHER_SUITES)
                         .overrideAuthority("cln") // the grpc plugin for core lightning does not know the domain, therefore it uses 'cln' by default. See https://docs.corelightning.org/docs/grpc.
                         .maxInboundMessageSize(RefConstants.MAX_GRPC_MESSAGE_SIZE)
                         .build();

--- a/app/src/main/java/app/michaelwuensch/bitbanana/backends/lnd/connection/LndConnection.java
+++ b/app/src/main/java/app/michaelwuensch/bitbanana/backends/lnd/connection/LndConnection.java
@@ -177,6 +177,7 @@ public class LndConnection {
                         .forAddress(host, port)
                         .hostnameVerifier(hostnameVerifier) // null = default hostnameVerifier
                         .sslSocketFactory(inspectingFactory) // null = default SSLSocketFactory
+                        .tlsConnectionSpec(RefConstants.TLS_VERSIONS, RefConstants.CIPHER_SUITES)
                         .maxInboundMessageSize(RefConstants.MAX_GRPC_MESSAGE_SIZE)
                         .build();
             }

--- a/app/src/main/java/app/michaelwuensch/bitbanana/util/RefConstants.java
+++ b/app/src/main/java/app/michaelwuensch/bitbanana/util/RefConstants.java
@@ -2,6 +2,9 @@ package app.michaelwuensch.bitbanana.util;
 
 import java.util.concurrent.TimeUnit;
 
+import okhttp3.CipherSuite;
+import okhttp3.TlsVersion;
+
 public class RefConstants {
 
     /* This value has to be increased if any changes are made, that break the current implementation.
@@ -113,4 +116,26 @@ public class RefConstants {
     // Other
     public static final String SETUP_MODE = "setupMode";
     public static final String BITBANANA_UTXO_LEASE_ID = "bb21212121212121212121212121212121212121212121212121212121212121";
+
+    // TLS versions to use for SSL connections
+    public static final String[] TLS_VERSIONS = {
+            TlsVersion.TLS_1_2.javaName(),
+            TlsVersion.TLS_1_3.javaName()
+    };
+    // Ciphers to use for SSL connections
+    public static final String[] CIPHER_SUITES = {
+            // TLS v1.3
+            CipherSuite.TLS_AES_128_GCM_SHA256.javaName(),
+            CipherSuite.TLS_AES_256_GCM_SHA384.javaName(),
+            CipherSuite.TLS_CHACHA20_POLY1305_SHA256.javaName(),
+            CipherSuite.TLS_AES_128_CCM_SHA256.javaName(),
+            CipherSuite.TLS_AES_128_CCM_8_SHA256.javaName(),
+            // TLS v1.2
+            CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256.javaName(),
+            CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256.javaName(),
+            CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384.javaName(),
+            CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384.javaName(),
+            CipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256.javaName(),
+            CipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256.javaName()
+    };
 }

